daml-lf/interpreter/src/test/scala/com/digitalasset/daml/lf/speedy/svalue/EqualitySpec.scala

```diff
@@ -217,55 +221,217 @@ class EqualitySpec extends WordSpec with Matchers with TableDrivenPropertyChecks
+      forEvery(positiveTestCases)((x, y) => areEqual(x, y))
+    }
+
+    "shortcut failure in variant" in {
+      val negativeTestCases = Table(
+        "first arg" -> "second arg",
+        u -> u,
+      )
+      val positiveTestCases = Table(
+        "first arg" -> "second arg",
+        u -> x1
+      )
+      forEvery(negativeTestCases) { (x, y) =>
+        an[SErrorCrash] should be thrownBy areEqual(x, y)
+        an[SErrorCrash] should be thrownBy areEqual(y, x)
+      }
+      forEvery(positiveTestCases) { (x, y) =>
+        areEqual(x, y)
+        areEqual(y, x)
+      }
+    }
+
+    "shortcut failure in list" in {
+      def list(xs: SValue*) = SList(FrontStack(xs))
+
+      val negativeTestCases = Table(
+        "first arg" -> "second arg",
+        list(u) -> list(u),
+        list(u, x1) -> list(u),
+        list(x1, u) -> list(x1, u),
+        list(x1, u, x1) -> list(x1, u),
+      )
+      val positiveTestCases = Table(
+        "first arg" -> "second arg",
+        list(x1, u) -> list(x2, u),
+        list(u) -> list(),
+        list() -> list(u)
+      )
+
+      forEvery(negativeTestCases) { (x, y) =>
+        an[SErrorCrash] should be thrownBy areEqual(x, y)
+        an[SErrorCrash] should be thrownBy areEqual(y, x)
+      }
+      forEvery(positiveTestCases) { (x, y) =>
+        areEqual(x, y)
+        areEqual(y, x)
+      }
+    }
+
+    "shortcut failure in optional" in {
+      val negativeTestCases = Table(
+        "first arg" -> "second arg",
+        SOptional(Some(u)) -> SOptional(Some(u)),
+      )
+      val positiveTestCases = Table(
+        "first arg" -> "second arg",
+        SOptional(None) -> SOptional(Some(u))
+      )
+      forEvery(negativeTestCases) { (x, y) =>
+        an[SErrorCrash] should be thrownBy areEqual(x, y)
+        an[SErrorCrash] should be thrownBy areEqual(y, x)
+      }
+      forEvery(positiveTestCases) { (x, y) =>
+        areEqual(x, y)
+        areEqual(y, x)
+      }
+    }
+
+    "shortcut failure in text map" in {
+      def map(xs: (String, SValue)*) = STextMap(HashMap(xs: _*))
+
+      val k1 = "a"
+      val k2 = "b"
+      assert("a" < "b")
+
+      val negativeTestCases = Table(
+        "first arg" -> "second arg",
+        map(k1 -> u) -> map(k1 -> u),
+        map(k1 -> u, k2 -> x1) -> map(k1 -> u),
+        map(k1 -> u, k2 -> x1) -> map(k1 -> u, k2 -> x2),
+        map(k1 -> x1, k2 -> u) -> map(k1 -> x1, k2 -> u),
+      )
+      val positiveTestCases = Table(
+        "first arg" -> "second arg",
+        map(k1 -> x1, k2 -> u) -> map(k1 -> x2, k2 -> u),
+        map(k1 -> x1, k2 -> u) -> map(k2 -> u),
+        map(k1 -> u) -> map(),
+      )
+      forEvery(negativeTestCases) { (x, y) =>
+        an[SErrorCrash] should be thrownBy areEqual(x, y)
+        an[SErrorCrash] should be thrownBy areEqual(y, x)
+      }
+      forEvery(positiveTestCases) { (x, y) =>
+        areEqual(x, y)
+        areEqual(y, x)
       }
     }
+
+    "shortcut failure in gen map" in {
```

Comment:
We don't have test cases where the comparison of the keys fails here, do we? If not, we should add some.

Reply:
I added some test cases.